### PR TITLE
stark: multi-task uncertainty weighting (Kendall 2018) — auto-balance surface/volume/τ losses

### PR DESCRIPTION
**Pull request by morganmcg1**
# Multi-Task Uncertainty Weighting (Kendall et al. 2018)

## Hypothesis

The yi training loss currently is a fixed-weight sum:
```
L = surface_loss_weight * L_surface + volume_loss_weight * L_volume
   + wallshear_y_weight * L_ws_y + wallshear_z_weight * L_ws_z
   + ...
```

Manually tuning these weights is brittle, and existing sweeps (PR #66 thorfinn, PR #309 thorfinn, PR #425 stark) have tested only a tiny region of the weight space. Kendall, Gal & Cipolla (2018) "Multi-Task Learning Using Uncertainty to Weigh Losses" (CVPR 2018) propose making the per-task weights **learnable log-variance parameters**:

```
L = sum_i [ exp(-s_i) * L_i + s_i ]
```

where `s_i = log(sigma_i^2)` is a learnable scalar per task. The model auto-balances tasks: high-noise tasks get down-weighted, low-noise tasks get up-weighted, **without any hand-tuning**. The `+ s_i` term prevents the trivial collapse `s_i → ∞`.

This is well-suited to DrivAerML because the per-task error scales differ by 2× (surface_p ≈ 5.5%, ws ≈ 10.5%, vol_p ≈ 13.5%) — fixed weights tuned for one error regime under-serve the others.

## Current Baseline

| Metric | Yi baseline (PR #309) |
|--------|----------------------|
| val_abupt (primary) | **9.039%** |
| test_abupt | ~10.2% |
| surface_pressure | ~5.5% |
| wall_shear_vector | ~10.5% |
| volume_pressure | ~13.5% |

**Active merge bar: val_abupt < 9.039%**.

### Step 1 — Add learnable log-variance parameters

In `train.py`, in the model class (or as a separate `nn.Module` registered on the model), add 4 learnable scalars:

```python
# Initialize log-variances to zero (uniform weighting at start, matches current baseline)
self.log_var_surface = nn.Parameter(torch.zeros(1))
self.log_var_volume = nn.Parameter(torch.zeros(1))
self.log_var_ws_y = nn.Parameter(torch.zeros(1))
self.log_var_ws_z = nn.Parameter(torch.zeros(1))
```

These must be registered on the **model** (so they go through Lion + DDP correctly), not as standalone state.

### Step 2 — Modify the training loss

Find the loss aggregation block in `train.py`. Currently it computes something like:
```python
loss = (config.surface_loss_weight * L_surface
       + config.volume_loss_weight * L_volume
       + config.wallshear_y_weight * L_ws_y
       + config.wallshear_z_weight * L_ws_z)
```

Replace with:
```python
if config.use_uncertainty_weighting:
    s_surf, s_vol, s_y, s_z = (model.log_var_surface, model.log_var_volume,
                               model.log_var_ws_y, model.log_var_ws_z)
    loss = (torch.exp(-s_surf) * L_surface + s_surf
          + torch.exp(-s_vol) * L_volume + s_vol
          + torch.exp(-s_y)   * L_ws_y    + s_y
          + torch.exp(-s_z)   * L_ws_z    + s_z).squeeze()
else:
    loss = <original fixed-weight sum>
```

Add a CLI flag `--use-uncertainty-weighting` (default False).

If the existing loss does not separately expose L_ws_y / L_ws_z (i.e., they are folded into L_surface), then keep the surface/volume two-task split:
```python
loss = torch.exp(-s_surf) * L_surface + s_surf + torch.exp(-s_vol) * L_volume + s_vol
```
and document this choice in the PR.

### Step 3 — Log the learned variances

Each step (or every N steps), log `exp(-s_i)` for each task to W&B as `loss_weight/<task>`. This is critical — without it we cannot diagnose whether the run drove a task to zero weight.

### Step 4 — Single-arm full-budget DDP

```bash
torchrun --standalone --nproc_per_node=4 train.py \
  --agent stark \
  --wandb-group stark-r30-uncertainty-weighting \
  --wandb-name "stark-uncertainty-arm-a" \
  --use-uncertainty-weighting \
  --optimizer lion \
  --lr 1e-4 --weight-decay 1e-4 \
  --no-compile-model --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-epochs 1 --clip-grad-norm 0.5 \
  --epochs 50
```

### Step 5 — Reporting

| Metric | Yi baseline | Arm A (uncertainty) | Δ |
|---|---|---|---|
| val_abupt | 9.039% | ?? | ?? |
| test_abupt | ~10.2% | ?? | ?? |
| surface_p (test) | 5.5% | ?? | ?? |
| ws_y (test) | 10.5% | ?? | ?? |
| ws_z (test) | 11.8% | ?? | ?? |
| vol_p (test) | 13.5% | ?? | ?? |

Plus the **final learned weights**: `exp(-s_surf)`, `exp(-s_vol)`, `exp(-s_ws_y)`, `exp(-s_ws_z)`. If any single weight has collapsed to <0.01 or grown to >100, that task has been effectively switched off — diagnose.

## What success looks like

- val_abupt < 9.039% → MERGE
- Final learned weights show a non-trivial pattern (e.g. volume up-weighted, surface down-weighted) → mechanism validated even if metric is null
- Loss diverges or one task collapses → either (a) the log-var grad path needs detaching from `s_i` term, or (b) the residual scales need normalising before applying — investigate and report

## Risk / fail-safe

- The Kendall paper notes one pathological case where regression-with-variance can drive `s_i → ∞`. Mitigations: clip `s_i` to `[-3, 3]` (effective weight in [exp(-3), exp(3)] = [0.05, 20]) or initialize at zero (already specified). If a clip is needed, add it explicitly.
- If implementing 4 separate ws_y/ws_z log-vars is messy, fall back to surface/volume 2-task version and document.
- Mark this PR ready for review when training completes and the per-task weight evolution is logged.

## Reference

- Kendall, Gal, Cipolla (2018) — "Multi-Task Learning Using Uncertainty to Weigh Losses for Scene Geometry and Semantics" — CVPR 2018.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #492 because ADVISOR: Closing — stark is not an active yi student (no pod on yi branch). The yi roster is the 16 students with active deployments: alphonse, askeladd, chihiro, edward, emma, fern, frieren, gilbert, haku, kohaku, nezuko, norman, senku, tanjiro, thorfinn, violet. The uncertainty-weighting hypothesis (Kendall 2018) is preserved in the research log and may be reassigned to an idle yi student if one becomes available..
